### ISSUE TITLE
Remote changes to property rules not taken into account

### ISSUE THREAD

**Issue by myniva** [akamai/terraform-provider-akamai#305]
### Terraform Version
```
Terraform v1.1.7
on linux_amd64
+ provider registry.terraform.io/akamai/akamai v1.12.1
```

### Affected Resource(s)
- `akamai_property`

### Terraform Configuration Files
```hcl
resource "akamai_property" "this" {
  name = var.domainname

  hostnames {
    cname_from             = var.domainname
    cname_to               = akamai_edge_hostname.this.edge_hostname
    cert_provisioning_type = "CPS_MANAGED"
  }

  rule_format = var.property_config_rule_format
  rules       = var.property_config_rules

  product_id  = local.product_id
  contract_id = local.contract_id
  group_id    = local.group_id
}
```

### Expected Behavior
If the rules of a property get changed outside of Terraform (via Console or PAPI or any other way that is possible), I expect Terraform to detect those changes on the next `apply` run. The latest version of the property should be updated to what is declared in the `akamai_property` Terraform resource.

### Actual Behavior
Terraform correctly identifies that there is a new version of the property. However, it does not revert the changes made manually so that the declared state matches the actual property rules tree again. The plan output looks like this:

```
Terraform detected the following changes made outside of Terraform since the last "terraform apply":

  # module.domain.akamai_property.this has changed
  ~ resource "akamai_property" "this" {
        id                 = "prp_123456"
      ~ latest_version     = 3 -> 4
        name               = "domain.com"
        # (12 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }
```

As seen above, Terraform **does** detect the newly created version, however it **DOES NOT** not propose to change the version rule tree. As a result of this, there is a configuration diff between what is defined in Terraform and what is configured on the Akamai property.

### Steps to Reproduce

1. Run `terraform apply` to create the Akamai property. There is no need to activate the current version.
2. Create a new version of the property in the console. Make a small modification and save the result.
3. Run `terraform apply` again.

**Comment by robertolopezlopez** [comment]
Thanks for opening this issue, @myniva .

I will try to reproduce from my side and hopefully find a fix or workaround. Could you please share with me the `product_id` which you are using? Thanks.

Regards,

Roberto

**Comment by myniva** [comment]
Thanks for looking into it. The `product_id` we're using is `prd_Site_Accel`.

**Comment by robertolopezlopez** [comment]
Hi,

Thank you for the extra details.

I am doing my testing with the following pieces:

    Terraform v1.0.5
    on darwin_amd64
    + provider registry.terraform.io/akamai/akamai v1.12.1

These are my steps:

1. Create property from terraform:
    ![image](https://user-images.githubusercontent.com/1519467/163181266-94b27598-ee95-4a3b-a721-e26f85535c43.png)
3. Change rule in property manager.
4. `terraform apply`
    ![image](https://user-images.githubusercontent.com/1519467/163182670-fe23c2fb-abdf-4e66-af86-3b0af54396d0.png)
5. Added an activation from terraform side (apparently your property is active, given that it is already v3) + `terraform apply`:
    ![image](https://user-images.githubusercontent.com/1519467/163184861-c6276aea-d1a8-478d-9495-d66959e9e1bc.png)
6. Edit in property manager to change some rule. Edit new version as v1 is no longer editable, then save. We have v2.
     ![image](https://user-images.githubusercontent.com/1519467/163185709-2ff5b578-cf34-4001-be9f-84168bb39f5d.png)
7. `terraform apply` - no rule changes shown, so no problem
    ![image](https://user-images.githubusercontent.com/1519467/163186083-82d34e11-5ed0-419b-a363-36276076d483.png)
   The property activation is overwritten, it takes some time to do the modification but after that everything seems correct. 
    ![image](https://user-images.githubusercontent.com/1519467/163186918-1a83046a-f3af-498d-ab0e-e8c81bb9db86.png)
    ![image](https://user-images.githubusercontent.com/1519467/163187124-6ef54531-651d-4ddf-915a-dc455a169617.png)

Have I missed anything in my workflow? For the moment I cannot see anything wrong in the provider behavior.

Regards,

Roberto

**Comment by robertolopezlopez** [comment]
I have used the following config for my test. Please compare with yours and let me know if we have missed any detail.

```hcl
terraform {
  required_providers {
    akamai = {
      source = "akamai/akamai"
    }
  }
  required_version = ">= 0.13.0"
}

provider "akamai" {
  edgerc = "~/.edgerc"
  config_section = "rlopezlo_cloud-sqa-shared"
}

resource "akamai_edge_hostname" "terraform-demo" {
  contract_id = "ctr_1-xxxxx"
  group_id = "grp_xxxxx"
  product_id = "prd_Web_App_Accel"
edge_hostname = "gh-i305-ii.com.edgesuite.net"
ip_behavior = "IPV4"
}

resource "akamai_property" "akaproperty" {
depends_on = [ akamai_edge_hostname.terraform-demo ]
    contract_id        = "ctr_1-xxxxx"
    group_id           = "grp_xxxxx"
    name               = "example-gh_i305-ii"
    product_id         = "prd_Web_App_Accel"
    rule_format        = "latest"

	hostnames {
		cert_provisioning_type = "CPS_MANAGED"
		cname_from             = "gh-i305-ii.com"
		cname_to               = akamai_edge_hostname.terraform-demo.edge_hostname
		cname_type             = "EDGE_HOSTNAME"

		cert_status {}
	    }
    rules              = jsonencode(
        {
            rules = {
                behaviors = [
                    {
                        name    = "origin"
                        options = {
                            cacheKeyHostname          = "ORIGIN_HOSTNAME"
                            compress                  = true
                            enableTrueClientIp        = true
                            forwardHostHeader         = "REQUEST_HOST_HEADER"
                            hostname                  = "ghi305-ii.com"
                            httpPort                  = 80
                            httpsPort                 = 443
                            originCertificate         = ""
                            originSni                 = true
                            originType                = "CUSTOMER"
                            ports                     = ""
                            trueClientIpClientSetting = false
                            trueClientIpHeader        = "True-Client-IP"
                            verificationMode          = "PLATFORM_SETTINGS"
                        }
                    },
                    {
                        name    = "cpCode"
                        options = {
                            value = {
                                cpCodeLimits = {
                                    currentCapacity = 589
                                    limit           = 3574
                                    limitType       = "account"
                                }
                                createdDate  = 1649842883000
                                description  = "example-gh.i305-ii"
                                id           = 1035619
                                name         = "example-gh.i305-ii"
                                products     = [
                                    "Web_App_Accel",
                                ]
                            }
                        }
                    },
                    {
                        name    = "caching"
                        options = {
                            behavior = "NO_STORE"
                        }
                    },
                    {
                        name    = "sureRoute"
                        options = {
                            enableCustomKey     = false
                            enabled             = true
                            forceSslForward     = false
                            raceStatTtl         = "30m"
                            srDownloadLinkTitle = ""
                            testObjectUrl       = "/akamai/testobject.html"
                            toHostStatus        = "INCOMING_HH"
                            type                = "PERFORMANCE"
                        }
                    },
                    {
                        name    = "tieredDistribution"
                        options = {
                            enabled               = true
                            tieredDistributionMap = "CH2"
                        }
                    },
                    {
                        name    = "prefetch"
                        options = {
                            enabled = true
                        }
                    },
                    {
                        name    = "allowPost"
                        options = {
                            allowWithoutContentLength = false
                            enabled                   = true
                        }
                    },
                    {
                        name    = "report"
                        options = {
                            logAcceptLanguage = false
                            logCookies        = "OFF"
                            logCustomLogField = false
                            logHost           = false
                            logReferer        = false
                            logUserAgent      = true
                        }
                    },
                    {
                        name    = "mPulse"
                        options = {
                            apiKey         = ""
                            bufferSize     = ""
                            configOverride = ""
                            enabled        = true
                            loaderVersion  = "V12"
                            requirePci     = false
                            titleOptional  = ""
                        }
                    },
                    {
                        name    = "allHttpInCacheHierarchy"
                        options = {
                            enabled = true
                        }
                    },
                ]
                children  = [
                    {
                        behaviors           = [
                            {
                                name    = "gzipResponse"
                                options = {
                                    behavior = "ALWAYS"
                                }
                            },
                        ]
                        criteria            = [
                            {
                                name    = "contentType"
                                options = {
                                    matchCaseSensitive = false
                                    matchOperator      = "IS_ONE_OF"
                                    matchWildcard      = true
                                    values             = [
                                        "text/*",
                                        "application/javascript",
                                        "application/x-javascript",
                                        "application/x-javascript*",
                                        "application/json",
                                        "application/x-json",
                                        "application/*+json",
                                        "application/*+xml",
                                        "application/text",
                                        "application/vnd.microsoft.icon",
                                        "application/vnd-ms-fontobject",
                                        "application/x-font-ttf",
                                        "application/x-font-opentype",
                                        "application/x-font-truetype",
                                        "application/xmlfont/eot",
                                        "application/xml",
                                        "font/opentype",
                                        "font/otf",
                                        "font/eot",
                                        "image/svg+xml",
                                        "image/vnd.microsoft.icon",
                                    ]
                                }
                            },
                        ]
                        criteriaMustSatisfy = "all"
                        name                = "Content Compression"
                        options             = {}
                    },
                    {
                        behaviors           = [
                            {
                                name    = "caching"
                                options = {
                                    behavior       = "MAX_AGE"
                                    mustRevalidate = false
                                    ttl            = "1d"
                                }
                            },
                            {
                                name    = "prefetch"
                                options = {
                                    enabled = false
                                }
                            },
                            {
                                name    = "prefetchable"
                                options = {
                                    enabled = true
                                }
                            },
                        ]
                        criteria            = [
                            {
                                name    = "fileExtension"
                                options = {
                                    matchCaseSensitive = false
                                    matchOperator      = "IS_ONE_OF"
                                    values             = [
                                        "aif",
                                        "aiff",
                                        "au",
                                        "avi",
                                        "bin",
                                        "bmp",
                                        "cab",
                                        "carb",
                                        "cct",
                                        "cdf",
                                        "class",
                                        "css",
                                        "doc",
                                        "dcr",
                                        "dtd",
                                        "exe",
                                        "flv",
                                        "gcf",
                                        "gff",
                                        "gif",
                                        "grv",
                                        "hdml",
                                        "hqx",
                                        "ico",
                                        "ini",
                                        "jpeg",
                                        "jpg",
                                        "js",
                                        "mov",
                                        "mp3",
                                        "nc",
                                        "pct",
                                        "pdf",
                                        "png",
                                        "ppc",
                                        "pws",
                                        "swa",
                                        "swf",
                                        "txt",
                                        "vbs",
                                        "w32",
                                        "wav",
                                        "wbmp",
                                        "wml",
                                        "wmlc",
                                        "wmls",
                                        "wmlsc",
                                        "xsd",
                                        "zip",
                                        "webp",
                                        "jxr",
                                        "hdp",
                                        "wdp",
                                        "pict",
                                        "tif",
                                        "tiff",
                                        "mid",
                                        "midi",
                                        "ttf",
                                        "eot",
                                        "woff",
                                        "woff2",
                                        "otf",
                                        "svg",
                                        "svgz",
                                        "webp",
                                        "jxr",
                                        "jar",
                                        "jp2",
                                    ]
                                }
                            },
                        ]
                        criteriaMustSatisfy = "all"
                        name                = "Static Content"
                        options             = {}
                    },
                    {
                        behaviors           = [
                            {
                                name    = "downstreamCache"
                                options = {
                                    behavior = "TUNNEL_ORIGIN"
                                }
                            },
                        ]
                        criteria            = [
                            {
                                name    = "cacheability"
                                options = {
                                    matchOperator = "IS_NOT"
                                    value         = "CACHEABLE"
                                }
                            },
                        ]
                        criteriaMustSatisfy = "all"
                        name                = "Dynamic Content"
                        options             = {}
                    },
                ]
                name      = "default"
                options   = {}
            }
        }
    )
}

resource "akamai_property_activation" "akaproperty_staging" {
     property_id = akamai_property.akaproperty.id
     contact  = ["example@example.com"]
     version  = akamai_property.akaproperty.latest_version
     # not specifying network will target STAGING
     note = "Sample activation"
}
```

**Comment by myniva** [comment]
Hi @robertolopezlopez 

Thanks for providing the exact Terraform configuration. The results on my side look very similar to what you have reported in in your previous comment. However, I don't agree with your statement in step 6 (_"no rule changes shown, so no problem"_):

The fact that there is no rule change shown **is** the actual problem. By performing what Terraform suggests, you would now enable V2 of the config on the STAGING env. But since V2 differs from the config that is declared in Terraform, we would now have a configuration drift.

Beside proposing to activate V2 on STAGING, the Terraform provider must also propose to modify V2 so that it matches the rules defined in Terraform. Of course, the modification has to take place before the activation (but I think that should already be the case due to the dependency tree).

**Comment by robertolopezlopez** [comment]
Hi @myniva ,

Thanks for the hint. So, according to your comment, this is actually the problem in my step 6:

![image](https://user-images.githubusercontent.com/1519467/163382421-102cefdf-72b4-4195-93ba-999f4ca09a0c.png)

... and the provider should detect not just the version change, but also the rule change from server side. Is that correct?

Roberto

**Comment by myniva** [comment]
Yes, that is correct 👍🏻

**Comment by robertolopezlopez** [comment]
OK, working on it in our internal repository. I will let you know about any updates

**Comment by robertolopezlopez** [comment]
Hi @myniva ,

I am thinking over and over your use case. Basically, you would like in step 6 that the provider would perform the following steps:

1. Refresh the local schema with the changes from server: this is done as initial step in the `ReadContext` operation of the `akamai_property` resource.
2. Ignore any diff between the refreshed local schema rules and configuration rules (in the resource `CustomizeDiff`).
3. Do not update the server rules with the local rules configuration (ignored in step 2).

I am close to say that step 2 is not possible to accomplish with the terraform-plugin-sdk v2, because when we get to the  `CustomizeDiff` function just two things are visible: the old schema state (from the server, set in refresh stage) and the new schema state (from the local `.tf` configuration, which for terraform `serves as the target source of truth for what exists on the backend API` [*](https://www.terraform.io/plugin/sdkv2/best-practices/detecting-drift)). There is no distinction at all for terraform at this point if the old copy comes from server refresh or not (read [here](https://www.terraform.io/plugin/sdkv2/resources/customizing-differences), [here](https://github.com/hashicorp/terraform-plugin-sdk/blob/v2.7.0/helper/schema/resource.go#L209) and [here](https://github.com/hashicorp/terraform-plugin-sdk/blob/2387aca0b8a1347c17b36b826c0f56d9258df457/helper/schema/resource.go#L314)). Any extra customization at this point would lead us once again to the same problem: we do not know if we are executing a clean update (without server rule changes) or your given scenario.

This is just my current idea of this problem, maybe I will still find some hack (unlikely).

Extra thoughts: would `terraform apply -refresh-only` help in your use case?

**Comment by robertolopezlopez** [comment]
The big problem here is that this property was created with terraform, then modified from server side (postman / property manager), then refreshed state from terraform side. This is not an use case on which terraform excels.

You may be able to sync those server changes with `terraform apply -refresh-only` and just after continue your work. Or create a property in server and then use our cli-terraform tool to import it without much issues.

But apart of that, mixing terraform with UI changes is not a good idea.

**Comment by robertolopezlopez** [comment]
After talking to my colleague from support I think that my last comments need a little bit more explanation.

We understand that it is difficult to change the way users work. In this case, it is probably difficult for the final user to do all changes from terraform. But, given the way terraform resources work, it is also difficult to change this default behavior. Let me just explain this `terraform apply -refresh-only` workaround.

So, I will go back to my previous property [from some comments above](https://github.com/akamai/terraform-provider-akamai/issues/305#issuecomment-1098035789).

This is what terraform would try to do next time we execute `terraform apply`:

1. Refresh local schema with remote changes (in this case, I disabled compression from the property rules).
![image](https://user-images.githubusercontent.com/1519467/163992531-65799300-b7e0-44d2-be16-032abfa4d3e5.png)
2. And then, try to overwrite these changes done from UI with the `.tf` configuration.
![image](https://user-images.githubusercontent.com/1519467/163992907-61b155a0-a1f8-4113-a579-7179a1424620.png)

I can understand that this is not good :-)

As I explained in my previous comments, it is difficult / impossible to completely fix this problem from the resource. But from the user side, this is easily fixable by executing `terraform apply -refresh-only` before `terraform apply`.

1. `terraform apply -refresh-only`: the local schema will get refreshed and the remote state will remain intact.
![image](https://user-images.githubusercontent.com/1519467/163993507-e0fe9507-3255-4c95-b96c-01dd1be73cbe.png)
2. `terraform apply`: as the local schema matches the remote state, terraform clearly informs the user that we are going to attempt an update with the old changes. At this point, the user can cancel the command and fix the `.tf` configuration.
![image](https://user-images.githubusercontent.com/1519467/163993942-e7b72f18-b0ed-460e-889f-896751ce8e30.png)

Is this good enough for you? Thank you for your feedback.

Roberto

**Comment by piotrpio** [comment]
Hello @myniva ,

The rules on the latest version are not being overwritten by changes in HCL because of invalid setting of `read_version` attribute in `resourcePropertyRead`. The `read_version` attribute is used in order to enable users to do `terraform import` on a specific version and then start working from there, but we should not even be setting it (fallback to latest) if the property is created from terraform rather then imported. 

I've already tested a simple PoC locally and it seems that the fix will be relatively simple, although for sure we will not be able to squeeze it into the current release, so it will have to wait until the next release.

**Comment by robertolopezlopez** [comment]
Hi @myniva ,

This should be fixed with our [v2.1.0 release](https://github.com/akamai/terraform-provider-akamai/releases/tag/v2.1.0). Please let us know in the (unlikely) case that this fix does not completely solve your problem :-)

Kind regards,

Roberto
